agent/config/conditional.go

```diff
@@ -24,17 +24,61 @@ const (
 	_ Conditional = iota
 	ExplicitlyEnabled
 	ExplicitlyDisabled
-	DefaultEnabled
+	NotSet
 )
 
+type BooleanDefaultTrue struct {
+	Value Conditional
+}
+
 // Enabled is a convenience function for when consumers don't care if the value is implicit or explicit
-func (b Conditional) Enabled() bool {
-	return b == ExplicitlyEnabled || b == DefaultEnabled
+func (b BooleanDefaultTrue) Enabled() bool {
+	return b.Value == ExplicitlyEnabled || b.Value == NotSet
+}
+
+// MarshalJSON is used to serialize the type to json, per the Marshaller interface
+func (b BooleanDefaultTrue) MarshalJSON() ([]byte, error) {
+	switch b.Value {
+	case ExplicitlyEnabled:
+		return json.Marshal(true)
+	case ExplicitlyDisabled:
+		return json.Marshal(false)
+	default:
+		return json.Marshal(nil)
+	}
+}
+
+// UnmarshalJSON is used to deserialize json types into Conditional, per the Unmarshaller interface
+func (b *BooleanDefaultTrue) UnmarshalJSON(jsonData []byte) error {
+	jsonString := string(jsonData)
+	jsonBool, err := strconv.ParseBool(jsonString)
+	if err != nil && jsonString != "null" {
+		return err
+	}
+
+	if jsonString == "" || jsonString == "null" {
+		b.Value = NotSet
+	} else if jsonBool {
+		b.Value = ExplicitlyEnabled
+	} else {
+		b.Value = ExplicitlyDisabled
+	}
+
+	return nil
+}
```

Comment:
Is it possible to restructure this code to 


```
       jsonString := string(jsonData)
       if jsonString == "" || jsonString == "null" {
		b.Value = NotSet
                return nil
        }
	jsonBool, err := strconv.ParseBool(jsonString)
	if err != nil  {
		return err
	}

	if jsonBool {
		b.Value = ExplicitlyEnabled
	} else {
		b.Value = ExplicitlyDisabled
	}

```
Just from readability point of view, in the current form it is bit confusing to understand what is happening.

Reply:
I think I prefer to keep it as is (it was what was there before), and to me it makes sense to group the error handling and b.value setting separate.